libraries/standard/http/include/http_client.h

```diff
@@ -0,0 +1,339 @@
+    HTTP_SECURITY_ALERT_PARSER_INVALID_CHARACTER,
+    HTTP_SECURITY_ALERT_INVALID_CONTENT_LENGTH,
+    /* TODO: Add return codes as implementation continues. */
+} HTTPStatus_t;
+
+/**
+ * @brief Request headers buffer and how many headers exist in the buffer.
+ */
+typedef struct HTTPRequestHeaders
+{
+    
+    /**
+     * @brief Buffer to hold the raw HTTP request headers.
+     * 
+     * This buffer is not available to modify until #HTTPNetworkInterface_t.recv
+     * for the response is invoked.
+     * 
+     * If this buffer is shared with the response, then this buffer cannot be
+     * modified until #HTTPClient_Send() returns.
+     */
+    uint8_t* pBuffer;
+    size_t bufferLen; /**< The length pBuffer. */
+    
+    /**
+     * @brief The actual amount of headers in the buffer, in bytes. This field
+     * is updated by the HTTP Client library functions #HTTPClient_AddHeader,
+     * and #HTTPClient_InitializeRequestHeaders.
+     */
+    size_t headersLen;
+} HTTPRequestHeaders_t;
+
+/**
+ * @brief Configurations of the initial request headers.
+ */
+typedef struct HTTPRequestInfo
+{
+    /**
+     * @brief The HTTP request method i.e. "GET", "POST", "PUT", or "HEAD". 
+     */
+    const char* method;
+    size_t methodLen; /**< The length of the method in bytes. */
+    
+    /**
+     * @brief The Request-URI to the objects of interest, i.e. "/path/to/item.txt".
+     */
+    const char* pPath;
+    size_t pathLen; /**< The length of the path in bytes. */
+
+    /**
+     * @brief The DNS resolvable name of the host or server, i.e. 
+     * "s3.amazonaws.com"
+     * 
+     * The host does not have a "https://" or "http://" prepending.
+     */
+    const char* pHost;
+    size_t hostLen; /**< The length of the host in bytes. */
+   
+    /**
+     * @brief Flags to activate other request header configurations. 
+     */
+    uint32_t flags;
+} HTTPRequestInfo_t;
+
+/**
+ * @brief Callback to intercept headers during the first parse through of the 
+ * response as it is received from the network.
+ */
+typedef struct httpResponseParsingCallback {
+    /**
+     * Invoked when both a header field and its associated header value is found.
+     * @param[in] pContext User context.
+     * @param[in] fieldLoc Pointer to the first char of the header field name in the response buffer.
+     * @param[in] fieldLen Length in bytes of the field name.
+     * @param[in] valueLoc Pointer to the first char of the header value in the response buffer.
+     * @param[in] valueLen Length in bytes of the value.
+     * @param[in] statusCode The HTTP response status-code.
+     */
+    void (*onHeaderCallback)( void* pContext, const char* fieldLoc, size_t fieldLen, const char* valueLoc, size_t valueLen, uint16_t statusCode );
+
+    /* Private context for the application. */
+    void *pContext;
+} httpclientHeaderParsingCallback_t;
+
+typedef struct HTTPResponse
+{
+    /**
+     * @brief Buffer for both the raw HTTP header and body. 
+     * 
```

Comment:
Please add similar information here about memory ownership model of the API. "The buffer memory is owned by the library when <HTTPClientSend> or <HTTPClientReadHeader> is called until the call returns. It should not be modified during the execution of the call."